### DISCUSSION TITLE
RuntimeTypeHandle typed typeof.

### DISCUSSION THREAD

**Discussion by MichalPetryka** [dotnet/csharplang#5473] [Language Ideas]
To obtain a `RuntimeTypeHandle`, one could use the `ldtoken` IL instruction, however C# currently doesn't expose the raw instruction today. It's only indirectly accessible through the `typeof` expression, which emits not only that, but also a call to `System.Type.GetTypeFromHandle`. This forces the users to do `typeof(T).TypeHandle`, which has a measurable performance impact as noted [here](https://github.com/dotnet/runtime/issues/5973).

I'd like to propose adding this:
```C#
RuntimeTypeHandle type = typeof(T);
```
The compiler would automatically decide whether the expression would return a `Type` or a `RuntimeTypeHandle` depending on where the result is passed to, like it currently does with `stackalloc` span and pointer. The compiler would omit the `call` and only emit the `ldtoken` in such case.

**Comment by FiniteReality** [comment]
One excellent use-case for this would be source generators which need some form of runtime type information, but don't want to resort to reflection. For example, I currently use the type information to query additional metadata about a given type, in scenarios where attributes on the type may not work directly, such as lazily initialised information. This information is passed between source-generated code, so I can't easily use source generators to add additional attributes.

Though personally, I'd like there to be a `methodof` or similar which surfaced a RuntimeMethodHandle too.

**Reply by koszeggy** [reply]
Maybe a bit old (before the `nameof` era) but may worth reading it: https://docs.microsoft.com/en-us/archive/blogs/ericlippert/in-foof-we-trust-a-dialogue

**Reply by FiniteReality** [reply]
Yeah, I'm aware of that post, and I'm aware of the issues, but one can dream :stuck_out_tongue:

**Comment by jkotas** [comment]
> This forces the users to do `typeof(T).TypeHandle`, which has a measurable performance impact as noted here.

`typeof(T)` and `typeof(T).TypeHandle` produce identical code in recent .NET runtimes. See 
https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIGYACMhgYQYG8aHunGAVAJ4AHGAwCy5ABQBKDlx4LiAdgYZhMCADNJAZQxQAlgDsA5tIDc87gF8rvBgCUArkYwH8MQSIAS2IwBMAG1ExUhk5agVFFTURLV19YzMAOi8YXwDgy0ieW2prIA

The issue that you have linked talks about performance overhead of `typeof(T).TypeHandle.Value` that is a different problem, and that you proposal would not address.
